otherlibs/unix/fork.c

```diff
@@ -19,14 +19,21 @@
 #include <caml/debugger.h>
 #include <caml/eventlog.h>
 #include "unixsupport.h"
+#include <caml/domain.h>
+#include <caml/fail.h>
 
 CAMLprim value unix_fork(value unit)
 {
   int ret;
+  if (caml_domain_is_multicore()) {
+    caml_failwith
+      ("Unix.fork may not be called while other domains were created");
```

Comment:
We almost support the case where other domains were created, then shut down, and domain 0 is back to being the only domain running. We'd just have to shut down domain 0's backup thread in this case. Is there a use case for that?

Reply:
I would expect this "burst of parallel activity that spawns many domains and joins them" pattern to be rather common in applications were some parallelism was added after-the-fact, in particular any application using parallelism only using (a multicore port of) parmap.

From the perspective of `fork`, lifting the restriction in the way you suggest would mean that applications could use both `fork` and `parmap`, as long as `fork` is never called from within the parallelized `parmap` computations. That sounds like a desirable property to me. If I remember correctly, the "is this really multicore?" check is used in other parts of the runtime, where the refined check could be similarly beneficial.

On the other hand, some parts use `Domain.at_first_spawn`, which runs user-provided logic when we first "go parallel" (in Format: switch to a less-efficient, buffered implementation of standard i/o formatters), but no way to "go back" if we were to become sequential again. I wonder if we should consider a more expressive combinator with a callback to call on a multicore->sequential transition. 

Pros:
- more expressive
- may be important for performance in practice, when the "sequential" state is noticeably faster

Cons:
- more complex
- harder to write two callbacks / to check that they work as expected)
- currently `at_first_spawn` is only used in Format, which probably does not justify the effort

Reply:
Can we delay this discussion to after the merge?  The current code is conservative w.r.t. `fork`, but that's perhaps for the best.